### PR TITLE
Expand integration test

### PR DESCRIPTION
**Pull request by craffit**
### What does this PR do?

- Add integration test to see whether agent/processs agent runs
- Add test to see whether connection data is passed

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: test/molecule-role/molecule/default/prepare.yml [hunk 1]
```diff
@@ -92,29 +96,52 @@
...
         name: stackstate-agent
         state: started
         daemon_reload: yes
+- name: Prepare Agent 1 listening
+  hosts: agent1
+  gather_facts: false
+  tasks:
+    - name: Open listening port on agent 1
+      shell: nohup nc -l -p 125 > /dev/null &
+      become: true
```
Comment (TizianoPerrucci) [inline test/molecule-role/molecule/default/prepare.yml]: if we use a higher port than 125 you don't need `become`, right ?

File: test/molecule-role/molecule/default/tests/test_agent.py [hunk 2]
```diff
@@ -8,11 +10,75 @@
 def test_stackstate_agent_is_installed(host):
     agent = host.package("stackstate-agent")
     assert agent.is_installed
-    # TODO
+    print agent.version
+    # TODO: Why is the verison prefixed by 1?
```
Comment (TizianoPerrucci) [inline test/molecule-role/molecule/default/tests/test_agent.py]: good question, should not be the case

File: test/molecule-role/molecule/default/tests/test_receiver.py [hunk 3]
```diff
@@ -13,3 +15,35 @@ def test_etc_docker_directory(host):
 def test_docker_compose_file(host):
     f = host.file('/home/ubuntu/docker-compose.yml')
     assert f.is_file
+
```
Comment (TizianoPerrucci) [inline test/molecule-role/molecule/default/tests/test_receiver.py]: should we check if the 3 containers we expect are running ?
and/or check the receiver/health to be OK ?

File: test/molecule-role/molecule/default/tests/util.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import time
+
+def wait_until(somepredicate, timeout, period=0.25, *args, **kwargs):
+  mustend = time.time() + timeout
+  while True:
+    try:
+        somepredicate(*args, **kwargs)
+        return
```
Comment (TizianoPerrucci) [inline test/molecule-role/molecule/default/tests/util.py]: you mean `return somepredicate()` ?
or does this do something magic ?

File: test/molecule-role/molecule/default/create.yml [hunk 5]
```diff
@@ -18,6 +20,10 @@
         from_port: 7077  # StackState Receiver port
         to_port: 7077
         cidr_ip: '0.0.0.0/0'
+      - proto: tcp
+        from_port: 125  # StackState Receiver port
```
Comment (TizianoPerrucci) [inline test/molecule-role/molecule/default/create.yml]: comment is misleading here, this is `directional connections test port`

File: test/molecule-role/docker-compose.yml [hunk 6]
```diff
@@ -10,11 +10,21 @@ services:
...
+  topic-api:
+    image: "508573134510.dkr.ecr.eu-west-1.amazonaws.com/stackstate-topic-api:master"
+    command: "-J-Xms128M -J-Xmx1G -J-XX:+ExitOnOutOfMemoryError -DconsoleLogging=true"
+    ports:
+      - 7070:7070
+    depends_on:
+      - kafka
+    environment:
+      KAFKA_BROKERS: kafka:9092
+      ZOOKEEPER_QUORUM: kafka
```
Comment (TizianoPerrucci) [inline test/molecule-role/docker-compose.yml]: why is this needed to the topic-api but not for the receiver ?

File: test/molecule-role/molecule/default/create.yml (hunk 5 above)
Comment (TizianoPerrucci) [reply test/molecule-role/molecule/default/create.yml]: and maybe we should make an ansible variable out of it, considered that is used in multiple places

File: test/molecule-role/docker-compose.yml (hunk 6 above)
Comment (craffit) [reply test/molecule-role/docker-compose.yml]: The topic api uses zookeeper to learn about what's going on in kafka (what topics are there). The receiver only produces its own stuff so does not need this global overview

File: test/molecule-role/molecule/default/prepare.yml (hunk 1 above)
Comment (craffit) [reply test/molecule-role/molecule/default/prepare.yml]: Higher ports could be already occupied,  so i'd rather use a low one

File: test/molecule-role/molecule/default/create.yml (hunk 5 above)
Comment (craffit) [reply test/molecule-role/molecule/default/create.yml]: Good catch, I'll do this

File: test/molecule-role/molecule/default/tests/test_receiver.py (hunk 3 above)
Comment (craffit) [reply test/molecule-role/molecule/default/tests/test_receiver.py]: We could,  i would say molecule is in charge of the machines,  so i'd not second guess that. 
Receiver health is also not that important to me, we test the integration, not the operating of the receiver

File: test/molecule-role/molecule/default/tests/util.py (hunk 4 above)
Comment (craffit) [reply test/molecule-role/molecule/default/tests/util.py]: I'll rename it. The predicate is not expected to return a value

File: test/molecule-role/molecule/default/tests/test_agent.py (hunk 2 above)
Comment (craffit) [reply test/molecule-role/molecule/default/tests/test_agent.py]: fixed

File: test/molecule-role/molecule/default/create.yml [hunk 7]
```diff
@@ -28,7 +34,7 @@
         to_port: 0
         cidr_ip: '0.0.0.0/0'
 
-    keypair_name: molecule_key
+    keypair_name: "molecule_key_{{ molecule_run_id }}"
```
Comment (TizianoPerrucci) [inline test/molecule-role/molecule/default/create.yml]: why do we need to override the key name?

File: test/molecule-role/molecule/default/tests/test_receiver.py [hunk 8]
```diff
@@ -13,3 +15,43 @@ def test_etc_docker_directory(host):
...
+        assert outgoing_conn["connectionType"] == "TCP"
+        incoming = next(record
+                        for record
+                        in json_data["messages"]
+                        if record["message"]["Connection"]
+                        ["localEndpoint"]["endpoint"]["port"] == conn_port
+                        )
+        incoming_conn = incoming["message"]["Connection"]
+        print incoming_conn
+        # TODO: Enable this once we use a .deb with directional relations
```
Comment (TizianoPerrucci) [inline test/molecule-role/molecule/default/tests/test_receiver.py]: spurious TODO ?

File: test/molecule.sh [hunk 9]
```diff
@@ -14,4 +20,6 @@ fi
 
 cd molecule-role
 
+echo =====MOLECULE_RUN_ID=${CI_JOB_ID}======AGENT_GITLAB_BRANCH=${CI_COMMIT_REF_NAME}===== $AWS_ACCESS_KEY_ID ==
```
Comment (TizianoPerrucci) [inline test/molecule.sh]: why do we need to print the `AWS_ACCESS_KEY_ID` ?
for security I would rather avoid if not essential, you can always check the settings in gitlab

File: test/molecule-role/molecule/default/create.yml (hunk 7 above)
Comment (Voronenko) [reply test/molecule-role/molecule/default/create.yml]: tool drops one during the run, if not present locally

Option: provide that key,  than can be reused

File: test/molecule-role/molecule/default/create.yml [hunk 10]
```diff
@@ -88,13 +97,15 @@
...
+    - debug: var="ec2_jobs"
+
     # Mandatory configuration for Molecule to function.
 
     - name: Populate instance config dict
       set_fact:
         instance_conf_dict: {
-          'instance': "{{ item.instances[0].tags.instance }}",
-          'address': "{{ item.instances[0].public_ip }}",
+          'instance': "{{ item.tagged_instances[0].tags.instance }}",
```
[the remark this answers is not in the document]
Comment (Voronenko) [reply test/molecule-role/molecule/default/create.yml]: In some cases instances[0] is not present, but almost direct copy in tagged_instances - is.

File: test/molecule-role/molecule/default/create.yml (hunk 7 above)
Comment (craffit) [reply test/molecule-role/molecule/default/create.yml]: i prefer the current solution. I'd rather not share keys. The key is witness of the fact the we are authorized for only for the current molecule_run_id
